tiledb/sm/query/writer.cc

```diff
@@ -858,67 +828,148 @@ Status Writer::compute_coord_dups(
 Status Writer::compute_coord_dups(std::set<uint64_t>* coord_dups) const {
   STATS_FUNC_IN(writer_compute_coord_dups_global);
 
-  auto coords_buff_it = attr_buffers_.find(constants::coords);
-  if (coords_buff_it == attr_buffers_.end())
+  if (coord_buffers_.empty()) {
     return LOG_STATUS(
         Status::WriterError("Cannot check for coordinate duplicates; "
                             "Coordinates buffer not found"));
+  }
 
-  auto coords_buff = (unsigned char*)coords_buff_it->second.buffer_;
-  auto coords_buff_size = *coords_buff_it->second.buffer_size_;
-  auto coords_size = array_schema_->coords_size();
-  auto coords_num = coords_buff_size / coords_size;
+  if (coords_num_ < 2)
+    return Status::Ok();
 
-  for (uint64_t i = 1; i < coords_num; ++i) {
-    if (!memcmp(
-            coords_buff + i * coords_size,
-            coords_buff + (i - 1) * coords_size,
-            coords_size)) {
+  // Prepare auxiliary vector for better performance
+  auto dim_num = array_schema_->dim_num();
+  std::vector<unsigned char*> buffs(dim_num);
+  for (unsigned d = 0; d < dim_num; ++d) {
+    const auto& dim_name = array_schema_->dimension(d)->name();
+    buffs[d] = (unsigned char*)coord_buffers_.find(dim_name)->second.buffer_;
+  }
+
+  std::mutex mtx;
+  auto statuses = parallel_for(1, coords_num_, [&](uint64_t i) {
+    // Check for duplicate in adjacent cells
+    bool found_dup = true;
+    for (unsigned d = 0; d < dim_num; ++d) {
+      if (memcmp(
+              buffs[d] + i * coord_sizes_[d],
+              buffs[d] + (i - 1) * coord_sizes_[d],
+              coord_sizes_[d]) != 0) {  // Not the same
+        found_dup = false;
+        break;
+      }
+    }
+
+    // Found duplicate
+    if (found_dup) {
+      std::lock_guard<std::mutex> lock(mtx);
       coord_dups->insert(i);
     }
+
+    return Status::Ok();
+  });
+
+  // Check all statuses
+  for (auto& st : statuses) {
+    if (!st.ok())
+      return st;
   }
 
   return Status::Ok();
 
   STATS_FUNC_OUT(writer_compute_coord_dups_global);
 }
 
+Status Writer::compute_coords_metadata(
+    const std::unordered_map<std::string, std::vector<Tile>>& tiles,
+    FragmentMetadata* meta) const {
+  auto coords_type = array_schema_->coords_type();
+  switch (coords_type) {
+    case Datatype::INT8:
+      return compute_coords_metadata<int8_t>(tiles, meta);
+    case Datatype::UINT8:
+      return compute_coords_metadata<uint8_t>(tiles, meta);
+    case Datatype::INT16:
+      return compute_coords_metadata<int16_t>(tiles, meta);
+    case Datatype::UINT16:
+      return compute_coords_metadata<uint16_t>(tiles, meta);
+    case Datatype::INT32:
+      return compute_coords_metadata<int>(tiles, meta);
```

Comment:
extreme nitpick: let's use `int32_t` here instead of `int`. same for `unsigned`->`uint32` below.